ui/lib/sync/addon/components/secrets/page/destinations/create-and-edit.ts

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+import { task } from 'ember-concurrency';
+import { waitFor } from '@ember/test-waiters';
+import { inject as service } from '@ember/service';
+import { findDestination } from 'vault/helpers/sync-destinations';
+import errorMessage from 'vault/utils/error-message';
+
+import type SyncDestinationModel from 'vault/models/sync/destination';
+import { ValidationMap } from 'vault/vault/app-types';
+import type FlashMessageService from 'vault/services/flash-messages';
+import type RouterService from '@ember/routing/router-service';
+import type StoreService from 'vault/services/store';
+
+interface Args {
+  destination: SyncDestinationModel;
+}
+
+export default class DestinationsCreateForm extends Component<Args> {
+  @service declare readonly flashMessages: FlashMessageService;
+  @service declare readonly router: RouterService;
+  @service declare readonly store: StoreService;
+
+  @tracked modelValidations: ValidationMap | null = null;
+  @tracked invalidFormMessage = '';
+  @tracked error = '';
+
+  get displayName() {
+    return findDestination(this.args.destination.type)?.displayName;
+  }
+
+  @task
+  @waitFor
+  *save(event: Event) {
+    event.preventDefault();
+
+    const { destination } = this.args;
+    const { isValid, state, invalidFormMessage } = destination.validate();
+
+    this.modelValidations = isValid ? null : state;
+    this.invalidFormMessage = isValid ? '' : invalidFormMessage;
+    if (!isValid) return;
+
+    try {
+      const verb = destination.isNew ? 'created' : 'updated';
+      yield destination.save();
+      this.flashMessages.success(`Successfully ${verb} the destination ${destination.name}`);
+
+      this.router.transitionTo(
+        'vault.cluster.sync.secrets.destinations.destination.details',
+        destination.type,
+        destination.name
+      );
+    } catch (error) {
+      this.error = errorMessage(error, 'Error saving destination. Please try again or contact support.');
+    }
+  }
+
+  @action
+  cancel() {
+    const method = this.args.destination.isNew ? 'unloadRecord' : 'rollbackAttributes';
```

Comment:
I don't think we're doing this anymore since `rollbackAttributes` handles unloading a new record. Per the docs:

> If the model `hasDirtyAttributes` this function will discard any unsaved changes. If the model `isNew` it will be removed from the store.

Reply:
I remember this discussion and just rollbackAttributes originally, but noticed the model wasn't cleaning up? I can test again to make sure I wasn't missing something

Reply:
This is the error I got when clicking `cancel` and I had `this.args.destination.rollbackAttributes()` The transition fails because the icon getter errors? However when I just unload the record everything is groovy? I'll hold off merging until I hear back from ya!
<img width="1263" alt="Screenshot 2023-10-24 at 5 18 44 PM" src="https://github.com/hashicorp/vault/assets/68122737/1fc5979a-f7b7-45e9-a25f-af5cfe365974">

Reply:
Hmm that's strange. Maybe somewhere is trying to access the unloaded model? Anyway, no harm in leaving it as is with the conditional `unloadRecord`.

Reply:
Yeah - I'm not sure. The "Cancel" button just takes you to the page where you select the destination type, which doesn't rely on any models

Reply:
I wonder if it does a rollback of the attributes first no matter what and then checks for `isNew` and then removes the record. I think that would account for this situation since `type` would been set to a value and then rolled back to `undefined` which triggered a recompute of the getter throwing the assertion.

Reply:
Aha! That must be it. I suppose I could add a conditional in the `icon()` getter to just return if there is no type

Reply:
I suppose the question should be if for some reason a different type is passed in do we want things to break? It is helpful in the sense that we can get immediate feedback as developers but I think we could add some handling within the routes to ensure that only the allowed types are attempting to be accessed and show an error otherwise instead.